sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/ByteBuddyDoFnInvokerFactory.java

```diff
@@ -332,40 +384,62 @@ public String subclass(TypeDescription.Generic superClass) {
   }
 
-  private static Implementation getRestrictionCoderDelegation(DoFnSignature signature) {
+  private static Implementation getRestrictionCoderDelegation(
+      TypeDescription doFnType, DoFnSignature signature) {
     if (signature.processElement().isSplittable()) {
       if (signature.getRestrictionCoder() == null) {
         return MethodDelegation.to(
             new DefaultRestrictionCoder(signature.getInitialRestriction().restrictionT()));
       } else {
         return new DowncastingParametersMethodDelegation(
+            doFnType,
             signature.getRestrictionCoder().targetMethod());
       }
     } else {
       return ExceptionMethod.throwing(UnsupportedOperationException.class);
     }
   }
 
-  private static Implementation splitRestrictionDelegation(DoFnSignature signature) {
+  private static Implementation splitRestrictionDelegation(
+      TypeDescription doFnType, DoFnSignature signature) {
     if (signature.splitRestriction() == null) {
       return MethodDelegation.to(DefaultSplitRestriction.class);
     } else {
-      return new DowncastingParametersMethodDelegation(signature.splitRestriction().targetMethod());
+      return new DowncastingParametersMethodDelegation(
+          doFnType, signature.splitRestriction().targetMethod());
     }
   }
 
   /** Delegates to the given method if available, or does nothing. */
-  private static Implementation delegateOrNoop(DoFnSignature.DoFnMethod method) {
+  private static Implementation delegateOrNoop(TypeDescription doFnType, DoFnSignature.DoFnMethod
+      method) {
     return (method == null)
         ? FixedValue.originType()
-        : new DoFnMethodDelegation(method.targetMethod());
+        : new DoFnMethodDelegation(doFnType, method.targetMethod());
   }
 
   /** Delegates to the given method if available, or throws UnsupportedOperationException. */
-  private static Implementation delegateWithDowncastOrThrow(DoFnSignature.DoFnMethod method) {
+  private static Implementation delegateWithDowncastOrThrow(
+      TypeDescription doFnType, DoFnSignature.DoFnMethod method) {
     return (method == null)
         ? ExceptionMethod.throwing(UnsupportedOperationException.class)
-        : new DowncastingParametersMethodDelegation(method.targetMethod());
+        : new DowncastingParametersMethodDelegation(doFnType, method.targetMethod());
+  }
+
+  /**
+   * Internal delegation class for generated {@link DoFnInvoker} instances.
+   *
+   * <p>This class should <i>not</i> be accessed directly, or by Beam users. It must be public for
+   * generated instances to have adequate access, as they are generated "inside" the invoked {@link
+   * DoFn} class.
+   */
+  public static class OnTimerMethodDelegation {
+    public static void invokeOnTimer(
+        @This DoFnInvokerBase invoker,
+        @Argument(0) String timerId,
+        @Argument(1) ExtraContextFactory args) {
+      ((OnTimerInvoker) invoker.onTimerInvokers.get(timerId)).invokeOnTimer(args);
```

Comment:
Couldn't we just define this in the `DoFnInvokerBase`? Is there a reason it needs to be done via bytecode? One valid response would be if you expected to do conditionals here later.